src/Assistant.ts

```diff
@@ -0,0 +1,387 @@
+import {
+  AssistantThreadsSetStatusResponse,
+  AssistantThreadsSetSuggestedPromptsResponse,
+  AssistantThreadsSetTitleResponse,
+  ChatPostMessageArguments,
+} from '@slack/web-api';
+import processMiddleware from './middleware/process';
+import {
+  AllMiddlewareArgs,
+  AnyMiddlewareArgs,
+  Context,
+  Middleware,
+  SayFn,
+  SlackEventMiddlewareArgs,
+} from './types';
+import { AssistantInitializationError, AssistantMissingPropertyError } from './errors';
+import {
+  AssistantThreadContext,
+  AssistantThreadContextStore,
+  DefaultThreadContextStore,
+  GetThreadContextFn,
+  SaveThreadContextFn,
+} from './AssistantThreadContextStore';
+
+/**
+ * Configuration object used to instantiate the Assistant
+ */
+export interface AssistantConfig {
+  threadContextStore?: AssistantThreadContextStore;
+  threadStarted: AssistantThreadStartedMiddleware | AssistantThreadStartedMiddleware[];
+  threadContextChanged?: AssistantThreadContextChangedMiddleware | AssistantThreadContextChangedMiddleware[];
```

Comment:
Just a comment for future maintainers; no need to change anything -- Allowing multiple listeners may cause confusion (e.g., unnecessarily calling the threadContextStore several times) and almost 100% developers need only one listener here. With that being said, perhaps we still should keep the consistency in bolt-js (while bolt-python/java allow only one listener by design).

Reply:
Good point. Adding more context: with this architecture and where it's been used with the two other features, we have never acted on behalf of the developer: what they provide is what we run. We deviate from this in this feature by automatically calling `save` on their behalf and behind the scenes, if certain conditions are not met.